### ISSUE TITLE
manifest key to enable automatic injection of content scripts after installation/update

### ISSUE THREAD

**Issue by tophf** [hober/web-extensions#617]
Benefits/goals:
* Fixes inconsistency (Chrome doesn't do it, Firefox and Safari do).
* Makes the extension immediately usable.
* Simplifies code: no need for a [nontrivial workaround](https://stackoverflow.com/a/11598753) that requires duplicating `matches` in `host_permissions` and doesn't support `exclude_matches` yet ([#123](https://github.com/w3c/webextensions/issues/123)).
* Doesn't require reloading the tabs, which may lose their state (games, modern infinite-scroll pages).
* Solves a common pain point for new developers so they don't waste a lot of time in debugger/stackoverflow/etc.

Possible solutions:
1. <s>`"run_at": ["document_start", "extension_installed", "extension_enabled"]` i.e. an alternative array syntax for `run_at`</s>
2. `"run_at_extension": ["installed", "enabled"]` i.e. an additional key
3. `"run_at_installed": true` and `"run_at_enabled": true`
4. `"run_at_installed_or_enabled": true`

Notes:
* In 1 and 2 the browser should disallow invalid/duplicate values.
* The solution should also apply to [#600](https://github.com/w3c/webextensions/issues/600) e.g. `"extension_invoked"` in 1, `"invoked"` in 2, `"run_at_invoked"` in 3 and 4.
* Eventually this should be the default mode, but for now the current default behavior should be explicitly documented:
   * [MDN `content_scripts`](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/content_scripts) should copy the install note from [MDN concepts](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Content_scripts#loading_content_scripts).
   * MDN could add a row in the compat table or add a note about Chromium's behavior.
   * [Chrome `content_scripts`](https://developer.chrome.com/docs/extensions/mv3/manifest/content_scripts/).

**Comment by carlosjeurissen** [comment]
Very much in favour of having this be aligned cross browsers! Safari also injects content scripts right after installation.

As for the possible solutions. More in favour of 2-4. As 1; using an array for "run_at" would not be backwards compatible. Mixing the two also seems confusing rather than clarifying.

**Comment by fregante** [comment]
This is also something I wrote some code for (https://github.com/fregante/webext-inject-on-install) but it requires the `tabs` and `scripting` permissions.

Related requests:

- injection timing: https://github.com/w3c/webextensions/issues/600
- event name: https://github.com/w3c/webextensions/issues/353

Here's what they could look like:

```diff
  {
    "content_scripts": [{
      "id": "main",
      "matches": ["*://*.example.com/*"],
      "scripts": ["content.js"],
      "run_at": "document_idle",
+     "run_on_active_tab": true,
+     "run_on_extension_enabled": true
    }]
  }
```

or

```diff
  {
    "content_scripts": [{
      "id": "main",
      "matches": ["*://*.example.com/*"],
      "scripts": ["content.js"],
      "run_at": "document_idle",
+     "run_on": ["active_tab", "extension_enabled"]
    }]
  }
```

or combine it with `run_at`


```diff
  {
    "content_scripts": [{
      "id": "main",
      "matches": ["*://*.example.com/*"],
      "scripts": ["content.js"],
+     "run_at": ["document_idle", "active_tab", "extension_enabled"]
    }]
  }
```

**Comment by carlosjeurissen** [comment]
In addition to injection. There is also the removal aspect in case content scripts get registered using registerContentScripts as mentioned in https://github.com/w3c/webextensions/issues/653.

So we might want to come up with a different naming which also includes the above mentioned behaviour of removing already injected styles (not applying to scripts as injected scripts can not be reversed).

**Comment by carlosjeurissen** [comment]
Discussed in during the Berlin meeting on 2025-03-28. A potential approach would be a new `can_run_later` property which would indicate if a content script with `run_at="document_start"`, can deal with being run later in the document. Thus skipping the reload dialog in Chrome. This could also be used as indication to inject content scripts after installation.

Setting this property to false could allow extensions to opt-in to the chrome behaviour after installation.

@oliverdunk is Chrome supportive?

**Comment by oliverdunk** [comment]
> [@oliverdunk](https://github.com/oliverdunk) is Chrome supportive?

Yes, we're supportive of that approach in general. I think this is definitely a case where it would be great to see a formal proposal as the exact meaning is likely to be nuanced.

**Comment by carlosjeurissen** [comment]
FYI, currently in Orion, content scripts are only injected directly on installation when the extension is detected to be a Firefox extension. So even if a Safari web extension is loaded in Orion, it currently follows the Chrome behaviour. See:
https://orionfeedback.org/d/12534-content-scripts-are-not-being-loaded-on-installation-for-safari-extensions

However when disabled extensions get re-enabled, Orion will not inject content scripts before a page refresh no-matter what browser-type extension is loaded.

**Comment by carlosjeurissen** [comment]
Permission changes are also handled differently. Currently Safari is the only browser which injects content scripts as soon as permissions change and when the extension gets re-enabled after being disabled before. A full table of the current behaviour is the following:

Browser | onInstalled | onUpdated | onEnabled | onPermission
-- | -- | -- | -- | --
chrome | ❌ | ❌ | ❌ | ❌
firefox | ✅ | ✅ | ✅ | ❌
safari | ✅ | ✅ | ✅ | ✅
orion-chrome | ❌ | ❌ | ❌ | ❌
orion-firefox | ✅ | ✅ | ❌ | ❌
orion-safari | ❌ | ❌ | ❌ | ❌

orion-safari means an extension is understood to be a Safari extension by Orion.

**Comment by dotproto** [comment]
@carlosjeurissen, can you describe (or link out to) testing procedures for the various Orion extension loading scenarios?

EDIT: I read back a bit more in this thread and over on the Orion feedback site. It sounds like the testing procedure is basically just installing an extension from CWS or AMO. I'm still not clear on how to install a Safari extension.

**Comment by carlosjeurissen** [comment]
@dotproto Installing a Safari extension on Orion can be achieved by picking the container app when loading extension from disk. It is not very obvious I agree.
